pallets/elections/src/impls.rs

```diff
@@ -293,6 +304,58 @@ where
 
         T::ValidatorRewardsHandler::add_rewards(rewards);
     }
+
+    fn calculate_underperforming_non_reserved_validators() {
+        let thresholds = CurrentEraCommitteeKickOutConfig::<T>::get();
+        let current_committee = T::SessionInfoProvider::current_committee();
+        let next_era_non_reserved = NextEraNonReservedValidators::<T>::get()
+            .into_iter()
+            .collect::<BTreeSet<_>>();
+        let non_reserved_nodes_in_next_era = current_committee.intersection(&next_era_non_reserved);
+        let expected_blocks_per_validator = Self::blocks_to_produce_per_session();
+        for validator in non_reserved_nodes_in_next_era {
+            let underperformance = match SessionValidatorBlockCount::<T>::try_get(&validator) {
+                Ok(block_count) => {
+                    let block_performance_ratio = match block_count <= expected_blocks_per_validator
+                    {
+                        true => Perbill::from_rational(block_count, expected_blocks_per_validator),
+                        false => Perbill::one(),
+                    };
+                    block_performance_ratio <= thresholds.minimal_expected_performance
+                }
+                Err(_) => true,
+            };
+            if underperformance {
+                Self::mark_validator_underperformance(&thresholds, validator);
+            }
+        }
+    }
+
+    fn mark_validator_underperformance(
+        thresholds: &CommitteeKickOutConfig,
+        validator: &T::AccountId,
+    ) {
+        let counter = UnderperformedValidatorSessionCount::<T>::mutate(&validator, |count| {
+            *count += 1;
+            *count
+        });
+        if counter >= thresholds.underperformed_session_count_threshold {
+            ToBeKickedOutFromCommittee::<T>::insert(
+                &validator,
+                KickOutReason::InsufficientUptime(counter),
+            );
+            UnderperformedValidatorSessionCount::<T>::remove(&validator);
+        }
+    }
+
+    fn clear_underperformance_session_counter(session: SessionIndex) {
+        let clean_session_counter_delay =
+            CurrentEraCommitteeKickOutConfig::<T>::get().clean_session_counter_delay;
+        if session % clean_session_counter_delay == 0 {
+            info!(target: "pallet_elections", "Clearing UnderperformedValidatorSessionCount");
+            let _result = UnderperformedValidatorSessionCount::<T>::clear(u32::MAX, None);
+        }
+    }
```

Comment:
Shouldn't this be set _per-validator_ ? Unless this is done this way on purpose.

Reply:
Do you mean that each validator needs to have its specific counter? Can be, but it requires additional `StorageMap`, so I don't think it is worth it given profits. This is too much corner case anyway imho, since we set this `clean_session_counter_delay` to quite long value anyway (960 sessions == 10 eras)

Reply:
Yes, I imagine validators will underperform with different "schedules" and some might be just on the verge of being kicked out but they luck out and the counter gets cleared, starts over and they continue to underperfrom without any consequences. In the worst case scenario, a validator would underperform 959+960 sessions in a row.

If this is done purposefully then that's fine, I can understand the explanation that it's overengineered without much practical benefit.

Reply:
I think the main point here is additional storage - rule of thumb is, we should not be adding more storage unless having strong reason to do so

Reply:
Another benefit of this being stateless is that changing the config does not require any changes to the counters, and thus it's overall simpler. In general, whenever you add some storage, then you make it more complicated to change the logic in the future, because you might then need to write migrations etc.